### PR TITLE
Improve the llm kv cache on fs and add the doc for apis

### PR DESCRIPTION
**Pull request by dashanji**
What do these changes do?
-------------------------
- Fixes the buffer overflow in hash computation.
- Refactor the i/o operation implementation.
- Parallelize the read/write I/O.
- Improve the python API of llm kv cache.
- Document the public APIs of llm kv cache.
- Bump up the manylinx image to 2024-03-25-9206bd9.
- Replace memcpy with concurrent_memcpy in blob storage.

Fixes https://github.com/v6d-io/v6d/issues/1793, https://github.com/v6d-io/v6d/issues/1832 and https://github.com/v6d-io/v6d/issues/1833.

### CODE REVIEW

File: docker/Dockerfile.vineyard-llm-python-dev [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+# generate final image
+FROM python:3.10-slim-buster
+
+SHELL ["/bin/bash", "-l", "-c"]
+COPY --from=wheel /work/fixed_wheels/*.whl /tmp/
+
+RUN cd /tmp &&\
+    pip3 install --no-cache-dir *.whl ipython
+
+ENTRYPOINT ["/bin/bash", "-l", "-c" ]
```
Comment (sighingnow) [inline docker/Dockerfile.vineyard-llm-python-dev:215]: Remove this file. I think current `vineyard-python-dev` would be enough as we no longer use the `renameat2` API.

Comment (dashanji) [reply docker/Dockerfile.vineyard-llm-python-dev:215]: Make sense. Thanks for pointing it out.

File: docker/pypa/Dockerfile.manylinux1 [hunk 2]
```diff
@@ -112,3 +112,20 @@ RUN echo "Installing apache-arrow ..." && \
...
+    ./bootstrap.sh && \
+    ./b2 install -j`nproc` link=static runtime-link=static variant=release threading=multi \
+        --with-atomic \
+        --with-chrono \
+        --with-date_time \
+        --with-filesystem \
+        --with-random \
+        --with-system \
+        --with-thread && \
+    rm -rf /deps
```
Comment (sighingnow) [inline docker/pypa/Dockerfile.manylinux1:131]: Remove this part... Boost shouldn't be a dependency for python bindings...

File: modules/llm-cache/ds/kv_state_cache_block.cc [hunk 3]
```diff
@@ -151,19 +151,22 @@ Status KVStateCacheBlockBuilder::Make(
...
+    LLMKV& keyState = kvState[currentLayer].first;
+    LLMKV& valueState = kvState[currentLayer].second;
+    memcpy(
+        keyState.data,
+        keyStateTensorBuilderList[currentLayer]->data() + index * tensorBytes,
+        tensorBytes);
+    memcpy(
+        valueState.data,
+        valueStateTensorBuilderList[currentLayer]->data() + index * tensorBytes,
+        tensorBytes);
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_block.cc:169]: Use `vineyard::memory::concurrent_memcpy`, defined in `common/memory/memcpy.h`.

File: modules/llm-cache/ds/kv_state_cache_block.cc [hunk 4]
```diff
@@ -190,19 +193,23 @@ bool KVStateCacheBlockBuilder::IsFull() {
...
-    LLMKV valueState = (kvState.find(currentLayer)->second).second;
+    LLMKV keyState = kvState[currentLayer].first;
+    LLMKV valueState = kvState[currentLayer].second;
     RETURN_ON_ASSERT((keyState.length == (size_t) this->tensorBytes &&
                       valueState.length == (size_t) this->tensorBytes));
 
     uint8_t* keyData = keyStateTensorBuilderList[currentLayer]->data();
     uint8_t* valueData = valueStateTensorBuilderList[currentLayer]->data();
+    LOG(INFO) << "dst:" << keyData + index * this->tensorBytes
+              << " src:" << keyState.data;
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_block.cc:212]: Looks like this line of debug log should be drop...

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 5]
```diff
@@ -82,43 +83,134 @@ Status KVStateCacheManager::Make(std::shared_ptr<KVStateCacheManager>& manager,
...
+ * manager.
+ *
+ * @param tokenList The token list to be updated.
+ * @param kvStateList The kv state list of the token list.
+ *                    It's a 2D vector, the first dimension is the token index,
+ *                    and the second dimension is the layer index.
+ *                    The kv state is a pair of LLMKV, the first is the K tensor
+ *                    and the second is the V tensor. It contains two fields:
+ *                    data and length. The data is the pointer to the tensor
+ * data, and the length is the size of the tensor data.
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:101]: alignment.

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 6]
```diff
@@ -82,43 +83,134 @@ Status KVStateCacheManager::Make(std::shared_ptr<KVStateCacheManager>& manager,
...
+ *
+ * @param tokenList The token list to be queried.
+ * @param kvStateList The kv state list of the token list.
+ *                    It must be initialized before calling this function,
+ * including the data and length of the kv tensor. It's a 2D vector, the first
+ * dimension is the token index, and the second dimension is the layer index.
+ *                    The kv state is a pair of LLMKV, the first is the K tensor
+ *                    and the second is the V tensor. It contains two fields:
+ *                    data and length. The data is the pointer to the tensor
+ * data, and the length is the size of the tensor data.
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:199]: alignment.

Comment (sighingnow) [reply modules/llm-cache/ds/kv_state_cache_manager.cc:199]: Add more content about if user should pre-allocate the buffers please.

Note that blobstroage and filestoreage requires different behaviour.

File: modules/llm-cache/ds/kv_state_cache_block.cc (hunk 3 above)
Comment (sighingnow) [reply modules/llm-cache/ds/kv_state_cache_block.cc:169]: Wait, why you need to add the memcpy to the blobstorage??

**Review by sighingnow — CHANGES_REQUESTED** [review]

File: modules/llm-cache/ds/kv_state_cache_block.cc (hunk 3 above)
Comment (dashanji) [reply modules/llm-cache/ds/kv_state_cache_block.cc:169]: I originally wanted the query API to be as consistent as filestorage, I will update it to the original one. Thanks.

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 6 above)
Comment (dashanji) [reply modules/llm-cache/ds/kv_state_cache_manager.cc:199]: image

@sighingnow Thanks for the advice. What do you think about such a description above?

**Comment by dashanji** [comment]
@sighingnow @vegetableysm Could you please take a look at this? Thanks.

File: modules/llm-cache/storage/local_file_storage.cc [hunk 7]
```diff
@@ -165,13 +160,24 @@ Status LocalFileStorage::Delete(std::string path) {
 
 std::string LocalFileStorage::GetTmpFileDir(std::string filePath) {
```
Comment (vegetableysm) [inline modules/llm-cache/storage/local_file_storage.cc:161]: Is ```std::string filePath``` still useful?

Comment (dashanji) [reply modules/llm-cache/storage/local_file_storage.cc:161]: Yes, it's the original file path computed by hash function.

File: modules/llm-cache/storage/file_storage.cc [hunk 8]
```diff
@@ -121,137 +222,226 @@ Status FileStorage::Update(
...
-  for (size_t i = 0; i < pathList.size(); i++) {
-    tokenLength = (i + 1) * batchSize;
+
+  std::vector<std::string> tempFilePaths(pathList.size());
+  auto fn = [&](size_t i) -> std::pair<int, Status> {
+    int tokenLength = (i + 1) * batchSize;
     std::shared_ptr<FileDescriptor> fd = CreateFileDescriptor();
-    std::string tmpPathStr = GetTmpFileDir(pathList[i]);
+    std::string tmpPathStr =
+        GetTmpFileDir(pathList[i]) + "-" + std::to_string(i);
```
Comment (vegetableysm) [inline modules/llm-cache/storage/file_storage.cc:287]: Perhaps this line should be followed by ```tempFilePaths[i] = tmpPathStr;```

File: modules/llm-cache/storage/local_file_storage.cc (hunk 7 above)
Comment (vegetableysm) [reply modules/llm-cache/storage/local_file_storage.cc:161]: But this param is not be used in this function.

File: modules/llm-cache/storage/file_storage.cc [hunk 9]
```diff
@@ -121,137 +222,226 @@ Status FileStorage::Update(
...
+  for (size_t i = 0; i < pathList.size(); i++) {
+    if (pathIndexMap.find(i) != pathIndexMap.end()) {
+      j += 1;
+    }
+  }
+  updated =
+      size_t(j * batchSize) < prefix.size() ? 0 : j * batchSize - prefix.size();
+  for (size_t i = j; i < pathList.size(); i++) {
+    VINEYARD_SUPPRESS(Delete(pathList[i]));
+    VINEYARD_SUPPRESS(Delete(tempFilePaths[i]));
```
[the remark this answers is not in the document]
Comment (vegetableysm) [reply modules/llm-cache/storage/file_storage.cc:376]: If so, delete the ```std::vector<std::string> tempFilePaths(pathList.size());``` and delete the related code.

File: modules/llm-cache/storage/file_storage.cc (hunk 8 above)
Comment (dashanji) [reply modules/llm-cache/storage/file_storage.cc:287]: Thanks for pointing out.

File: modules/llm-cache/storage/local_file_storage.h [hunk 10]
```diff
@@ -26,7 +26,7 @@ limitations under the License.
 namespace vineyard {
 
 struct LocalFileDescriptor : public FileDescriptor {
-  std::fstream fstream;
+  int fd;
   std::string path;
```
Comment (vegetableysm) [inline modules/llm-cache/storage/local_file_storage.h:30]: Please initialize this field in ```Open``` function.

File: modules/llm-cache/storage/local_file_storage.h [hunk 11]
```diff
@@ -26,7 +26,7 @@ limitations under the License.
...
   int lockFD;
```
Comment (vegetableysm) [inline modules/llm-cache/storage/local_file_storage.h:31]: If this field is no longer used, please remove it.

Comment (dashanji) [reply modules/llm-cache/storage/local_file_storage.h:31]: Ok.

File: modules/llm-cache/storage/local_file_storage.h (hunk 10 above)
Comment (dashanji) [reply modules/llm-cache/storage/local_file_storage.h:30]: Thanks.

File: modules/llm-cache/storage/local_file_storage.cc (hunk 7 above)
Comment (dashanji) [reply modules/llm-cache/storage/local_file_storage.cc]: Removed it.

File: python/vineyard/llm/tests/test_llm.py [hunk 12]
```diff
@@ -32,71 +33,134 @@ def test_kv_cache_update_and_query_on_blob(vineyard_ipc_sockets):
...
+    assert matched == len(tokens)
+
+    assert len(kv_tensors) == len(kv_tensors_from_cache)
+    for kv, kv_from_cache in zip(original_kv_tensors, kv_tensors_from_cache):
+        assert len(kv) == len(kv_from_cache)
+        for (k_tensor, v_tensor), (queried_k_tensor, queried_v_tensor) in zip(
+            kv, kv_from_cache
+        ):
+            assert torch.equal(k_tensor, queried_k_tensor)
+            assert torch.equal(v_tensor, queried_v_tensor)
```
Comment (sighingnow) [inline python/vineyard/llm/tests/test_llm.py:166]: Remove the usages of torch (as it is more harder to install). Numpy is enough for unittests of llm kvcache module.

**Review by sighingnow — CHANGES_REQUESTED** [review]

File: python/vineyard/llm/tests/test_llm.py (hunk 12 above)
Comment (dashanji) [reply python/vineyard/llm/tests/test_llm.py]: Ok.

**Comment by dashanji** [comment]
The I/O tests CI failed, maybe we can fix it in another one.

**Pull request merged**